pkg/handler/v2/external_cluster/node.go

```diff
@@ -84,52 +84,56 @@ func ListNodesMetricsEndpoint(userInfoGetter provider.UserInfoGetter, projectPro
+	}
+}
 
-		project, err := common.GetProject(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, req.ProjectID, &provider.ProjectGetOptions{IncludeUninitialized: false})
+func getClusterNodesMetrics(ctx context.Context, userInfoGetter provider.UserInfoGetter, projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider, clusterProvider provider.ExternalClusterProvider, privilegedClusterProvider provider.PrivilegedExternalClusterProvider, projectID, clusterID string) ([]apiv1.NodeMetric, error) {
+	nodeMetrics := make([]apiv1.NodeMetric, 0)
+
+	project, err := common.GetProject(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, projectID, &provider.ProjectGetOptions{IncludeUninitialized: false})
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+	cluster, err := getCluster(ctx, userInfoGetter, clusterProvider, privilegedClusterProvider, project.Name, clusterID)
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	isMetricServer, err := clusterProvider.IsMetricServerAvailable(cluster)
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	if isMetricServer {
+		client, err := clusterProvider.GetClient(cluster)
 		if err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
 		}
-		cluster, err := getCluster(ctx, userInfoGetter, clusterProvider, privilegedClusterProvider, project.Name, req.ClusterID)
-		if err != nil {
-			return nil, common.KubernetesErrorToHTTPError(err)
+		nodes := &corev1.NodeList{}
+		if err := client.List(ctx, nodes); err != nil {
+			return nil, err
+		}
+		availableResources := make(map[string]corev1.ResourceList)
+		for _, n := range nodes.Items {
+			availableResources[n.Name] = n.Status.Allocatable
 		}
 
-		isMetricServer, err := clusterProvider.IsMetricServerAvailable(cluster)
-		if err != nil {
+		nodeDeploymentNodesMetrics := make([]v1beta1.NodeMetrics, 0)
+		allNodeMetricsList := &v1beta1.NodeMetricsList{}
+		if err := client.List(ctx, allNodeMetricsList); err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
 		}
 
-		if isMetricServer {
-			client, err := clusterProvider.GetClient(cluster)
-			if err != nil {
-				return nil, common.KubernetesErrorToHTTPError(err)
-			}
-			nodes := &corev1.NodeList{}
-			if err := client.List(ctx, nodes); err != nil {
-				return nil, err
-			}
-			availableResources := make(map[string]corev1.ResourceList)
-			for _, n := range nodes.Items {
-				availableResources[n.Name] = n.Status.Allocatable
-			}
-
-			nodeDeploymentNodesMetrics := make([]v1beta1.NodeMetrics, 0)
-			allNodeMetricsList := &v1beta1.NodeMetricsList{}
-			if err := client.List(ctx, allNodeMetricsList); err != nil {
-				return nil, common.KubernetesErrorToHTTPError(err)
+		for _, m := range allNodeMetricsList.Items {
+			if _, ok := availableResources[m.Name]; ok {
```

Comment:
what is the need for this? I mean we are getting availableResources, metrics for the same pair of nodes, so why are we checking if availableResources has the node data

Reply:
availableResources is a map with available resources for the node
allNodeMetricsList := &v1beta1.NodeMetricsList{} - taken from metrics server
when we combine those two structures then we get the whole metrics structure for the node

This whole block is taken from `kubectl top node` implementation

Reply:
Yes I got that, but both availableResources and allNodeMetricsList are for same set of nodes, right?
then we are running a loop to check if availableResources has the nodes from allNodeMetricsList